File: PythonSolutions/Neetcode150/binarytree_problems/199_binary-tree-right-side-view.py
from typing import List
from collections import deque
# Definition for a binary tree node.
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right
class Solution:
    def rightSideView(self, root: TreeNode) -> List[int]:
        if not root:
            return []
        
        queue = deque([root])
        res = []

        while queue:
            n = len(queue)

            # Level order traversal
            for i in range(n):
                curr = queue.popleft()
                # In the current level, if we are at the last node add to result (gets right most value)
                if i == n - 1:
                    res.append(curr.val)
                
                if curr.left:
                    queue.append(curr.left)
                if curr.right:
                    queue.append(curr.right)
    
        return res

File: TypescriptSolutions/Neetcode150/binary-tree_problems/199_binary-tree-right-side-view.ts
export {};

import { Queue } from "@datastructures-js/queue";
class TreeNode {
    val: number
    left: TreeNode | null
    right: TreeNode | null
    constructor(val?: number, left?: TreeNode | null, right?: TreeNode | null) {
        this.val = (val===undefined ? 0 : val)
        this.left = (left===undefined ? null : left)
        this.right = (right===undefined ? null : right)
    }
}


const rightSideView = (root: TreeNode | null): number[] => {
    if (!root) return [];

    const res: number[] = [];
    const queue = new Queue([root]);

    while (!queue.isEmpty()) {
        const n = queue.size();

        for (let i = 0; i < n; i++) {
            const curr = queue.dequeue()!;

            if (i === n - 1) res.push(curr.val);

            if (curr.left) queue.enqueue(curr.left);
            if (curr.right) queue.enqueue(curr.right);
        }
    }

    return res;
};